# Part 1. Equation of a Slime

How many late days are you using for this assignment? 0

In [403]:
# Imports section
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score, KFold
from sklearn.preprocessing import PolynomialFeatures
from sklearn.pipeline import make_pipeline
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neural_network import MLPClassifier
import numpy as np

## 1. Loading the dataset

In [404]:
# Using pandas load the dataset
# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)

ds = pd.read_csv('science_data_large.csv')
print(ds.head(15))
ds.info()
print("Number of datapoints:", ds.shape[0])


    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 n

## 2. Splitting the dataset

In [405]:
# Take the pandas dataset and split it into our features (X) and label (y)

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
# For grading consistency use random_state=42 

features = ds[['Temperature °C', 'Mols KCL']]
label = ds[['Size nm^3']]

X_train, X_test, y_train, y_test, = train_test_split(features, label, test_size=0.1, random_state=42)

## 3. Perform a Linear Regression

In [406]:
# Use sklearn to train a model on the training set

# Create a sample datapoint and predict the output of that sample with the trained model

# Report the score for that model using the default score function property of the SKLearn model, in your own words (markdown, not code) explain what the score means

# Extract the coefficients and intercept from the model and write an equation for your h(x) using LaTeX

reg = LinearRegression().fit(X_train.values, y_train.values)
print(reg.predict([[1000, 100]]))
print(reg.score(X_train.values, y_train.values))
print(reg.coef_)
print(reg.intercept_)

[[560024.44043709]]
0.8610082946223788
[[ 866.14641337 1032.69506649]]
[-409391.47958341]


Write the linear equation of a slime: (example equation: $E = mc^2$)

$\text{size} = 866.14641337 \text{ (Temperature)} + 1032.69506649 \text{ (KCL)} - 409391.47958341$

Report on score and explain meaning: 
The score for this model is 0.8610082946223788. This means that the linear regression fits the model pretty well, with the two independent variables (features) predicting the dependent variable (label) fairly linearly.

## 4. Use Cross Validation

In [407]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
# For grading consistency use n_splits=5 and random_state=42
kf = KFold(n_splits=5, shuffle=True, random_state=42)
scores = cross_val_score(LinearRegression(), features.values, label.values, cv=kf, scoring='r2')
# Report on their finding and their significance
print(scores)
print(scores.mean())

[0.86151889 0.82742341 0.87195173 0.88166206 0.85609101]
0.8597294202684646


Write findings here:
The average score of a LinearRegression on the data was 0.8597294202684646, indicating that the linear regression fits the model data very well, with the variables exhibiting a highly linear relationship.

## 5. Using Polynomial Regression

In [408]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
# Perform k-fold cross validation (as above)
model = make_pipeline(PolynomialFeatures(degree=2), LinearRegression())

# Report on the metrics and output the resultant equation as you did in Part 3.

kf = KFold(n_splits=5, shuffle=True, random_state=42)
scores = cross_val_score(model, features.values, label.values, cv=kf, scoring='r2')
# Report on their finding and their significance
print(scores)
print(scores.mean())
model.fit(X_train.values, y_train.values)
linear_model = model.named_steps['linearregression']
poly_features = model.named_steps['polynomialfeatures']
coefficients = linear_model.coef_
intercept = linear_model.intercept_
print(coefficients)
print(intercept)


[1. 1. 1. 1. 1.]
1.0
[[ 0.00000000e+00  1.20000000e+01 -1.27197088e-07  1.26458843e-11
   2.00000000e+00  2.85714287e-02]]
[2.0479376e-05]


Write the polynomial equation of a slime: (example equation: $E = mc^2$)

$\text{size} = 12 \text{ (Temperature)} - 0.000000123113604 \text{ (KCL)} - 0.0000000000105648823\text{ (Temperature)}^2 + 2 \text{ (KCL)}^2 + 0.0285714287 \text{(KCL)}\times \text{(Temperature)} + 0.0000165727106$

Report on the score and interpret:
The score is 1 for each split. This indicates that the Polynomial Regression found by this model perfectly predicts the available data, with each data point lying on the curve.

# Part 2. Chronic Kidney Disease Prediction via Classification

Create code and markdown cells as needed to perform classification and report on your results

In [409]:
# Load the dataset. Then train and evaluate the classification models.

ckd = pd.read_csv('ckd_feature_subset.csv')
X = ckd[['age', 'bp', 'wbcc', 'appet_poor', 'appet_good', 'rbcc']]
y = ckd[['Target_ckd']]
kf = KFold(n_splits=5, shuffle=True, random_state=42)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)

#Logistic Regression
model = LogisticRegression()
model.fit(X_train.values, y_train.values.ravel())

scores = cross_val_score(model, X.values, y.values.ravel(), cv=kf, scoring='r2')
arr = np.array([scores.mean(), scores.std()])
print(arr)

#SVC
model = SVC(kernel='linear')
model.fit(X_train.values, y_train.values.ravel())

scores = cross_val_score(model, X.values, y.values.ravel(), cv=kf, scoring='r2')
arr = np.vstack((arr, [scores.mean(), scores.std()]))
print(arr)

#K-nearest
model = KNeighborsClassifier()
model.fit(X_train.values, y_train.values.ravel())

scores = cross_val_score(model, X.values, y.values.ravel(), cv=kf, scoring='r2')
arr = np.vstack((arr, [scores.mean(), scores.std()]))
print(arr)

[0.28389548 0.29254864]
[[0.28389548 0.29254864]
 [0.31520861 0.33873804]]
[[0.28389548 0.29254864]
 [0.31520861 0.33873804]
 [0.63212215 0.2676169 ]]


In [410]:
#Neural Network
model = MLPClassifier(hidden_layer_sizes=(10, 10), max_iter=2000, random_state=42)

model.fit(X_train.values, y_train.values.ravel())

scores = cross_val_score(model, X.values, y.values.ravel(), cv=kf, scoring='r2')
arr = np.vstack((arr, [scores.mean(), scores.std()]))
print(arr)

[[0.28389548 0.29254864]
 [0.31520861 0.33873804]
 [0.63212215 0.2676169 ]
 [0.68308645 0.17044065]]


## Results and Conclusion for Classification Experiments

The final table, with the mean and std of each classification in each row was: 
[[0.28389548 0.29254864]
 [0.31520861 0.33873804]
 [0.63212215 0.2676169 ]
 [0.68308645 0.17044065]]

From these results, we can conclude that the neural network, implemented using sklearn's MLPClassifier, performed the best, with the highest average $r^2$ score and the lowest standard deviation, indicating a consistently high performance. A clear second was the k-nearest neighbors classification, with only a 0.05 lower mean when compared to the neural network and the second lowest standard deviation. Logistic Regression and SVC both performed much worse, with half as high averages and much higher standard deviations, indicating large variability in performance. The neural network probably performed the best because it is the most complex of the 4 classification models, with the ability to recognize patterns in the data and change according to patterns. Neural networks also continually improve, with deeper layers adding onto the observations of shallower layers.

In [411]:
# Experiments with Neural Network.
model = MLPClassifier(hidden_layer_sizes=(10, 10), max_iter=2000, random_state=42)

model.fit(X_train.values, y_train.values.ravel())

scores = cross_val_score(model, X.values, y.values.ravel(), cv=kf, scoring='r2')
arr2 = np.array([scores.mean(), scores.std()])

#number 2: 
model = MLPClassifier(hidden_layer_sizes=(20, 20), max_iter=2000, random_state=42)

model.fit(X_train.values, y_train.values.ravel())

scores = cross_val_score(model, X.values, y.values.ravel(), cv=kf, scoring='r2')
arr2 = np.vstack((arr2, [scores.mean(), scores.std()]))

#number 3:
model = MLPClassifier(activation='tanh', hidden_layer_sizes=(10, 10), max_iter=2000, random_state=42)

model.fit(X_train.values, y_train.values.ravel())

scores = cross_val_score(model, X.values, y.values.ravel(), cv=kf, scoring='r2')
arr2 = np.vstack((arr2, [scores.mean(), scores.std()]))

#number 4
model = MLPClassifier(hidden_layer_sizes=(10, 10), max_iter=200, random_state=42)

model.fit(X_train.values, y_train.values.ravel())

scores = cross_val_score(model, X.values, y.values.ravel(), cv=kf, scoring='r2')
arr2 = np.vstack((arr2, [scores.mean(), scores.std()]))

#number 5:
model = MLPClassifier(hidden_layer_sizes=(3, 3), max_iter=3000, random_state=42)

model.fit(X_train.values, y_train.values.ravel())

scores = cross_val_score(model, X.values, y.values.ravel(), cv=kf, scoring='r2')
arr2 = np.vstack((arr2, [scores.mean(), scores.std()]))
print(arr2)

C:\Users\Main\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\Main\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\Main\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\Main\AppDa

[[0.68308645 0.17044065]
 [0.78119048 0.21901708]
 [0.8402381  0.16352203]
 [0.24222881 0.25995383]
 [0.71439959 0.20739829]]


## Results and Conclusion for Neural Network Experiments

The biggest difference for the classification was the activation parameter changing to 'tanh', the function used in the hidden layers of the network. This difference both increased the mean score by .14 and decreased the standard deviation by 0.01, meaning that the use of a hyperbolic tangent function correlated to an increase in performance. Both an increase and decrease in the hidden layer size corresponded to an increase of mean score, with the increase in hidden layer size about 3x as great as the decrease in hidden layer size. Finally, a massive decrease in the number of iterations drastically decreased the mean score and increased std, as the model was not able to fully converge.